libflux/src/semantic/analyze.rs

```diff
@@ -0,0 +1,2688 @@
+    lit: &'a ast::StringLiteral,
+    fresher: &mut Fresher,
+) -> Result<StringLiteral<'a>> {
+    Ok(StringLiteral {
+        loc: &lit.base.location,
+        type_info: type_info(fresher),
+        value: lit.value.clone(),
+    })
+}
+
+fn analyze_boolean_literal<'a>(
+    lit: &'a ast::BooleanLiteral,
+    fresher: &mut Fresher,
+) -> Result<BooleanLiteral<'a>> {
+    Ok(BooleanLiteral {
+        loc: &lit.base.location,
+        type_info: type_info(fresher),
+        value: lit.value,
+    })
+}
+
+fn analyze_float_literal<'a>(
+    lit: &'a ast::FloatLiteral,
+    fresher: &mut Fresher,
+) -> Result<FloatLiteral<'a>> {
+    Ok(FloatLiteral {
+        loc: &lit.base.location,
+        type_info: type_info(fresher),
+        value: lit.value,
+    })
+}
+
+fn analyze_integer_literal<'a>(
+    lit: &'a ast::IntegerLiteral,
+    fresher: &mut Fresher,
+) -> Result<IntegerLiteral<'a>> {
+    Ok(IntegerLiteral {
+        loc: &lit.base.location,
+        type_info: type_info(fresher),
+        value: lit.value,
+    })
+}
+
+fn analyze_unsigned_integer_literal<'a>(
+    lit: &'a ast::UnsignedIntegerLiteral,
+    fresher: &mut Fresher,
+) -> Result<UnsignedIntegerLiteral<'a>> {
+    Ok(UnsignedIntegerLiteral {
+        loc: &lit.base.location,
+        type_info: type_info(fresher),
+        value: lit.value,
+    })
+}
+
+fn analyze_regexp_literal<'a>(
+    lit: &'a ast::RegexpLiteral,
+    fresher: &mut Fresher,
+) -> Result<RegexpLiteral<'a>> {
+    Ok(RegexpLiteral {
+        loc: &lit.base.location,
+        type_info: type_info(fresher),
+        value: lit.value.clone(),
+    })
+}
+
+fn analyze_duration_literal<'a>(
+    lit: &'a ast::DurationLiteral,
+    fresher: &mut Fresher,
+) -> Result<DurationLiteral<'a>> {
+    Ok(DurationLiteral {
+        loc: &lit.base.location,
+        type_info: type_info(fresher),
+        value: convert_duration(&lit.values)?,
+    })
+}
+
+fn analyze_time_literal<'a>(
+    lit: &'a ast::DateTimeLiteral,
+    fresher: &mut Fresher,
+) -> Result<DateTimeLiteral<'a>> {
+    Ok(DateTimeLiteral {
+        loc: &lit.base.location,
+        type_info: type_info(fresher),
+        value: lit.value,
+    })
+}
+
+// In these tests we test the results of semantic analysis on some ASTs.
+// NOTE: we do not care about locations.
+// We create a default base node and clone it in various AST nodes to make the borrow checker happy.
+// We then use a pointer to its location in semantic nodes.
+// This works because standard equality between references checks the content of the pointers and not
+// the value of the pointers themselves.
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::semantic::types::{Fresher, Tvar};
+    use pretty_assertions::assert_eq;
+
+    // TestFresher mocks the fresher so that we don't have to test Tvar ids.
+    // That is not straightforward because the IncrementingFresher is used at different depths in
+    // the semantic graph.
+    struct TestFresher {}
```

Comment:
I believe the derivative crate will allow you to do something like this:
```Rust
#[derive(Derivative)]
#[derivative(PartialEq)]
pub struct Expression {
    ....
    #[derivative(PartialEq="ignore")]
    pub mono_type: MonoType,
}
```
so that you can ignore the monotype. Do we want to do that? We could ignore the source location as well.

On an "unrelated" note, I now see why you made `Fresher` a trait.

Reply:
Mmmmm, do we really want to make equality ignore the source location?

Two  identical identifiers, for instance, in two different positions are not the same identifier.

Even though I like this Rust directive, I wouldn't like to change the nodes' behavior for tests.

I am waiting for your feedback on this.
(I would really like to have some like Go's comparison options)

Reply:
Thinking more about it, the type can be excluded from the comparison, but I'd avoid to exclude the location for the reason above.

We could say that, in general, no two nodes in the same place should have a different type. So, we can safely exclude the type.